dashboard/signals.py

```diff
@@ -0,0 +1,33 @@
+"""
+Signals for user profiles
+"""
+from django.db.models.signals import post_save
+from django.dispatch import receiver
+
+from dashboard.models import CachedEnrollment, ProgramEnrollment
+
+
+@receiver(post_save, sender=CachedEnrollment, dispatch_uid="update_programenrollment")
+def create_programenrollment(sender, instance, created, **kwargs):  # pylint: disable=unused-argument
+    """
+    Signal handler to create Program enrollment when the CachedEnrollment table is updated
+    """
+    user = instance.user
+    program = instance.course_run.course.program
+    if instance.data is not None:
+        ProgramEnrollment.objects.get_or_create(
+            user=user,
+            program=program
+        )
+    else:
+        # check if there is any other entry with data != None
+        # if not delete the enrollment
+        active_enrollment_count = CachedEnrollment.objects.filter(
+            user=user,
+            course_run__course__program=program
+        ).exclude(data__isnull=True).count()
+        if active_enrollment_count == 0:
+            ProgramEnrollment.objects.filter(
+                user=user,
+                program=program
+            ).delete()
```

Comment:
what about doing this `else:` logic in a `pre_save` signal instead? something like:

```python
# might need to use something other than 'instance.pk' to let us know this is an update...
if not instance.data and instance.id is not None:
  instance_in_db = CachedEnrollment.objects.filter(id=instance.id).first()
  if instance_in_db.data:
    active_enrollment_count = CachedEnrollment.objects.filter(...)
    if active_enrollment_count == 0:
      ProgramEnrollment.objects.filter(
          user=user,
          program=program
      ).delete()
```

you can do the delete safely before the save since ProgramEnrollment has no formal relationship with the CachedEnrollment. this adds some extra code, and it still executes a query most times against `CachedEnrollment` but that query is only on the primary key so it's cheap. the main benefit is that this avoids excessive querying that would happen in the vast majority of cases